### ISSUE TITLE
Add overload to LoggerConfiguration.Wrap that allows caller to create the sink to be wrapped

### ISSUE THREAD

**Issue by augustoproiete** [serilog/serilog#1441]
Opening this one to start a conversation - happy to send a PR once (and if) we have a design agreed.

I'm building a sink that wraps other sinks, similar to what `Serilog.Sinks.Map` does, using `LoggerConfiguration.Wrap`. One of my requirements is to be able to catch any exceptions thrown by the sink (or sinks) being wrapped.

If my configuration has only one sink, `Wrap` returns an instance of that one sink I configured, which allows me to catch any exceptions when I call `Emit` on the wrapped sink. However, if my configuration has more than one sink, `Wrap` will aggregate them in a `SafeAggregateSink` which means I no longer can catch exceptions given `SafeAggregateSink` will catch them and write messages to `SelfLog`.

[LoggerSinkConfiguration.cs#L253-L255](https://github.com/serilog/serilog/blob/5e2164756efb6023a548b4eb91a71ed7edc3d9e1/src/Serilog/Configuration/LoggerSinkConfiguration.cs#L253-L255):
```
    var enclosed = sinksToWrap.Count == 1 ?
        sinksToWrap.Single() :
        new SafeAggregateSink(sinksToWrap);
```

I'd like to be able to tell `Wrap` which sink to use for aggregating multiple sinks, so that I can use my own implementation of an `AggregateSink` which would give me control over what to do in case of an exception.

With that said, I'd propose an overload that takes a function `Func<List<ILogEventSink>, ILogEventSink>` that replaces the code above and returns the right `ILogEventSink` to be wrapped.

Does that make sense? Is there an alternative that I'm missing?

ps: I did consider just bringing the code of `LoggerSinkConfiguration.Wrap` into my sink and make the changes, but it relies on `private` and `internal` members which my sink can't access.

**Comment by nblumhardt** [comment]
Hi @augustoproiete!

It sounds like what you want is very close to how a `LoggerAuditSinkConfiguration.Wrap()` method would work; I don't think the method exists, yet, but it would make sense for it to be exposed. Would this be a workable option?

**Comment by augustoproiete** [comment]
Hi @nblumhardt!

Yes, a `LoggerAuditSinkConfiguration.Wrap()` would indeed be very close. The only other thing is that I would prefer `AggregateSink` not to write the exceptions to `SelfLog` [as it does today](https://github.com/serilog/serilog/blob/master/src/Serilog/Core/Sinks/AggregateSink.cs#L44).

**Comment by nblumhardt** [comment]
Sorry, trying to get some clear time to consider the implications of the `SelfLog` change and auditing, just a heads-up to let you know this isn't forgotten!

I can't help feeling that, in an auditing situation, `AggregateSink` should probably just be transparent to exceptions and not try to continue writing to other sinks if one sink has failed.

I'd need to dig into the other usages of this class - any extra analysis appreciated, here!

**Comment by augustoproiete** [comment]
Thanks for following up, @nblumhardt.

My scenario is not really an auditing scenario, so the fact that `AggregateSink` attempts to write to all sinks, and throws an `AggregateException` at the end for the ones that failed actually works well for me.

In fact, I'd like to know more information about the issue - such as the sink name (or type) that failed, for instance. That's why I suggested an overload of `LoggerSinkConfiguration.Wrap` that would allow me to provide a `Func` to create the `AggregateSink` responsible for wrapping the multiple sinks, so that I can have full control over handling (or not) exceptions when writing to multiple sinks.

The current behavior of `LoggerSinkConfiguration.Wrap` is actually inconsistent comparing between one sink vs multiple sinks. For instance:

- If the user configures a _single_ sink, `LoggerSinkConfiguration.Wrap` provides the exact sink instance to wrap. The sink can throw exceptions, which bubble up and my wrapper can catch, and handle (or not).

- However, as you know, if the user configures more than one sink, `LoggerSinkConfiguration.Wrap` wraps them first into a `SafeAggregateSink` - which means exceptions no longer bubble up and my wrapper has not way of knowing if messages were emitted to all sinks or not - **that's the issue I'm trying to address**.

Is there a valid reason for this inconsistency? Exceptions raised by a single sink  bubble up and are caught anyway (and written to the `SelfLog`), so in theory `LoggerSinkConfiguration.Wrap` should be using `AggregateSink` instead of `SafeAggregateSink` - unless I'm missing something.

Here is a simple repro, with a `BadSink` that always throws an exception, and a `WrapperSink` that prints the type of the sink being wrapped, and the exception that occurred in the underlying sink:

```csharp
internal class BadSink : ILogEventSink
{
    public void Emit(LogEvent logEvent)
    {
        throw new NotImplementedException();
    }
}
```
```csharp
internal class WrapperSink : ILogEventSink
{
    private readonly ILogEventSink _innerSink;

    public WrapperSink(ILogEventSink innerSink)
    {
        _innerSink = innerSink ?? throw new ArgumentNullException(nameof(innerSink));
    }

    public void Emit(LogEvent logEvent)
    {
        Console.WriteLine(_innerSink.GetType().Name);

        try
        {
            _innerSink.Emit(logEvent);
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex);
        }
    }
}
```

```csharp
var loggerSinkConfiguration = new LoggerConfiguration().WriteTo;

static void Configure(LoggerSinkConfiguration wt) => wt.Sink<BadSink>();

ILogEventSink sink = null;
LoggerSinkConfiguration.Wrap(loggerSinkConfiguration,
    s => sink = new WrapperSink(s), Configure,
        LevelAlias.Minimum, null);

var log = new LoggerConfiguration().WriteTo.Sink(sink)
    .CreateLogger();

log.Information("Hey");
```

In the example above, the sink wrapped will be a `BadSink`, and exceptions will bubble up to the wrapper sink as expected.

Changing the configuration to write to more than one sink, proves that exceptions no longer bubble up, and the wrapper thinks that `Emit` succeeded in the underlying sink (which is not true - both sinks failed):

```csharp
static void Configure(LoggerSinkConfiguration wt) => wt
    .Sink<BadSink>()
    .WriteTo.Sink<BadSink>();
```

---

In summary, I think that:
- `LoggerSinkConfiguration.Wrap` shouldn't be wrapping multiple sinks with `SafeAggregateSink` and instead should be using `AggregateSink` by default
- `LoggerSinkConfiguration.Wrap` should provide a way for the caller to override this default and provide a `CustomAggregateSink` if needed.

Does that help frame the use-case? Sorry for the long text & code.

**Comment by nblumhardt** [comment]
I think there's a good case made for using `AggregateSink` rather than `SafeAggregateSink` inside `LoggerSinkConfiguration.Wrap()`. I'd need to think about the construction hook more carefully (hopefully some more time for that once Seq 2020.1 is out :-) ) - would the safe -> aggregate sink change be enough to unblock you in the meantime? PR would be welcome for that one, and would give us a good starting point to anchor the rest of the discussion.

**Comment by augustoproiete** [comment]
Great! Swapping `SafeAggregateSink` -> `AggregateSink` is a good step forward and will give wrappers a chance to handle exceptions.

I've just opened PR #1449 to make this tiny change. Happy to continue the discussion after you had some time to think more about the construction hook, and of course happy to send PRs for what comes out of that.

_ps: Good luck shipping Seq 2020.1!_

**Comment by skomis-mm** [comment]
Just for the reference..
I was trying to remember what was the context about the adding SafeAggregateSink. I think it comes from #1154

Briefly, it was about `Serilog.Sinks.Async` rework based on `Wrap(..)` method to make `BackgroundWrapperSink` as a lightweight decorator over underlying sink(s) sink instead of the old implementation based on the secondary ILogger.

The secondary logger on the surface exposes `SafeAggregatingSink` because it is configured with default Serilog's setup. 

So I was under assumption that the `Wrap` should preserve that to protect background thread from unhandled exceptions. But actually it is a unnecessary because it itself catch all of them.
